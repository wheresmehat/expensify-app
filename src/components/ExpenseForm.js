import React, { Component } from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

moment.locale("en-gb"); // breaks testing, uncomment for production

class ExpenseForm extends Component {

    state = {

        description: this.props.expense ? this.props.expense.description : "",
        amount: this.props.expense ? (this.props.expense.amount / 100).toString() : "",
        note: this.props.expense ? this.props.expense.note : "",
        createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
        calendarFocused: false,
        error: ""
    };

    onDescriptionChange = (event) => {
        
        const description = event.target.value;

        this.setState(() => ({ description }));
    }

    onNoteChange = (event) => {

        const note = event.target.value;

        this.setState(() => ({ note }));
    };

    onAmountChange = (event) => {

        const amount = event.target.value;
    
        if (!amount || /^\d+(?:\.\d{0,2})?$/.test(amount)) {
            
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = (createdAt) => {

        if (createdAt) {

            this.setState(() => ({ createdAt }));
        }
    };

    onFocusChange = ({ focused }) => {

        this.setState(() => ({ calendarFocused: focused }));
    };

    onSubmit = (event) => {

        event.preventDefault();

        if (!this.state.description || !this.state.amount) {

            this.setState(() => ({ error: "Please provide description and amount." }));
        }
        else {

            this.setState(() => ({ error: "" }));

            const expense = { 
                
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                note: this.state.note,
                createdAt: this.state.createdAt.valueOf()    
            };
            
            this.props.onSubmit(expense);
        }

    };

    render() {

        return (

            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        name="description" 
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker 
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false} 
                    />
                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>
                        Add Expense
                    </button>
                </form>
            </div>
        );
    }

};

export default ExpenseForm;