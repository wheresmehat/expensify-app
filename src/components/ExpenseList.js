import React from "react";
import { connect } from "react-redux";

import getVisibleExpenses from "../selectors/expenses";
import ExpenseListItem from "./ExpenseListItem";

export const ExpenseList = (props) => {     // export for testing purposes

    return (

        <div>

            {
                props.expenses.length === 0 ?
                    <p>No expenses</p>
                    :
                    props.expenses.map(expense => <ExpenseListItem key={expense.id} {...expense} />)
            }
        </div>
    );

};

const mapStateToProps = (state) => (

    { expenses: getVisibleExpenses(state.expenses, state.filters) }
);

export default connect(mapStateToProps)(ExpenseList);

//const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);
//export default ConnectedExpenseList;
