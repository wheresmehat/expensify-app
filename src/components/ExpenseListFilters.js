import React, { Component } from "react";       
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";

import * as filterActions from "../actions/filters";

export class ExpenseListFilters extends Component {

    state = {

        calendarFocused: null
    };

    onTextChange = (event) => { 
        
        this.props.setTextFilter(event.target.value) 
    };

    onSelectChange = (event) => {

        if (event.target.value === "date") {

            this.props.sortByDate();
        }
        else if (event.target.value === "amount") {

            this.props.sortByAmount();
        }

    };

    onDatesChange = ({ startDate, endDate }) => {

        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calendarFocused) => {

        this.setState(() => ({ calendarFocused }));
    };

    render() {

        return (

            <div>
                <input 
                    type="text" 
                    value={this.props.filters.text}
                    onChange={this.onTextChange} 
                />
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onSelectChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                />
            </div>
        );
    }

};

const mapStateToProps = (state) => {

    return { filters: state.filters };
};

export default connect(mapStateToProps, filterActions)(ExpenseListFilters);

/*

if we connected the component like so:

    export default connect(mapStateToProps)(ExpenseListFilter);

then on the this.props we have filters but we also have the dispatch function and we can call it inside the component with the action object:

    onChange={(event) => { this.props.dispatch({type: "SET_TEXT_FILTER", text: event.target.value}) }}

 or use the setTextFilter action creator to create the object:

    onChange={(event) => { this.props.dispatch(setTextFilter(event.target.value)) }}


BUT if we connect the component with the action creator, passing it as a second argument inside an object:

    export default connect(mapStateToProps, { setTextFilter })(ExpenseListFilter);

then on the this.props we have filters but we also have the setTextFilter action creator and we can use it inside the component without needing dispatch:

    onChange={(event) => { this.props.setTextFilter(event.target.value) }}

*/

/*

if we only need dispatch on the this.props we can connect the component like so:

    export default connect()(ExpenseListFilter);

*/

/* 

using the shorthand notation for mapDispatchToProps


we can write:

    EXAMPLE 1

    const mapDispatchToProps = (dispatch) => {

        return {

            setTextFilterProp: (text) => {  // we name the prop

                dispatch(setTextFilter(text))
            }
        };

    };

    export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);


or:

    EXAMPLE 2

    const mapDispatchToProps = (dispatch) => {

        return {

            setTextFilter: (text) => {  // it can be the same name as the action creator

                dispatch(setTextFilter(text))
            }
        };

    };

    export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);

and if the arguments passed to our prop are in the same order as the arguments passed to the action creator we can use an object shorthand and we don't need to use mapDispatchTothis.props:

    EXAMPLE 1

    export default connect(mapStateToProps, { setTextFilterProp: setTextFilter })(ExpenseListFilter);

if the names are the same then it's even shorter:

    EXAMPLE 2

    export default connect(mapStateToProps, { setTextFilter })(ExpenseListFilter);



// https://egghead.io/lessons/javascript-redux-using-mapdispatchtothis.props-shorthand-notation

*/