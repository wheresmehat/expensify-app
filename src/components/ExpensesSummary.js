import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import numeral from "numeral";

import getVisibleExpenses from "../selectors/expenses";
import getExpensesTotal from "../selectors/expenses-total";

export const ExpensesSummary = ({ expensesCount, expensesTotal }) => {

    const totalFormatted = numeral(expensesTotal / 100).format("$0,0.00");

    return (

        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">
                    Viewing <span>{expensesCount}</span> {expensesCount === 1 ? "expense" : "expenses"} totalling <span>{totalFormatted}</span>
                </h1>
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add Expense</Link>
                </div>
            </div>
        </div>
    );

};

const mapStateToProps = (state) => {

    const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);

    return { 
        
        expensesCount: visibleExpenses.length,
        expensesTotal: getExpensesTotal(visibleExpenses)
    };
}; 

export default connect(mapStateToProps)(ExpensesSummary);