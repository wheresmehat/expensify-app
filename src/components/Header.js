import React from "react";
import { NavLink } from "react-router-dom";
import { connect } from "react-redux";

import { startLogout } from "../actions/auth";

export const Header = ({ startLogout }) => (

    <header>
        <h1>Expensify</h1>
        <NavLink to="/" activeClassName="is-active" exact={true}>Dashboard</NavLink>
        <NavLink to="/create" activeClassName="is-active">Create expense</NavLink>
        <button
            onClick={startLogout}   
        >
            Logout
        </button>
    </header>

);

export default connect(null, { startLogout })(Header);