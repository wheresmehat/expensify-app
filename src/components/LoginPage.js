import React from "react";
import { connect } from "react-redux";

import { startLogin } from "../actions/auth";

export const LoginPage = ({ startLogin }) => {

    return (

        <div>
            <button
                onClick={startLogin}
            >
                Login
            </button>
        </div>
    );  
};

export default connect(null, { startLogin })(LoginPage);
