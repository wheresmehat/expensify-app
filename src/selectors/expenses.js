import moment from "moment";

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {

    return expenses.filter((expense) => {

        const createdAtMoment = moment(expense.createdAt);

        const textMatch = expense.description.toLowerCase().indexOf(text.toLowerCase()) > -1; // or use ES6 includes

        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, "day") : true;

        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, "day") : true;

        return textMatch && startDateMatch && endDateMatch;

    }).sort((firstExpense, secondExpense) => {

        if (sortBy === "date") {

            return secondExpense.createdAt - firstExpense.createdAt;
        }
        else if (sortBy === "amount") {

            return secondExpense.amount - firstExpense.amount;  
        }

    });

};

export default getVisibleExpenses;
