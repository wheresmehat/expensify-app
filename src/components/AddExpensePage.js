import React, { Component } from "react";
import { connect } from "react-redux";

import ExpenseForm from "./ExpenseForm";
import { addExpense } from "../actions/expenses";

export class AddExpensePage extends Component {     // export for testing purposes

    onSubmit = (expense) => {

        this.props.addExpense(expense);
        this.props.history.push("/");
    };

    render() {

        return (

            <div>
                <h1>Add Expense</h1>
                <ExpenseForm onSubmit={this.onSubmit} />
            </div>
        );
    }

};

export default connect(null, { addExpense })(AddExpensePage);


/*

const mapDispatchToProps = (dispatch) => {

    return {

        myAddExpense: (expense) => {   // we name the prop and use as props.myAddExpense(expense) above

            dispatch(addExpense(expense))   // addExpense is the action creator
        }
    };

};

export default connect(null, mapDispatchToProps)(AddExpensePage);


or without mapDispatchToProps if they have the same parameters in the same order:

export default connect(null, { myAddExpense: addExpense })(AddExpensePage);


or use the same name:

export default connect(null, { addExpense: addExpense })(AddExpensePage);


or using ES6 object shorthand for props and value of same name:

export default connect(null, { addExpense })(AddExpensePage);
*/