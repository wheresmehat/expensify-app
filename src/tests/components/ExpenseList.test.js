import React from "react";
import { shallow } from "enzyme";

import { ExpenseList } from "../../components/ExpenseList"; // import named export without connect
import expenses from "../fixtures/expenses";

test("should render ExpenseList with expenses", () => {

    const wrapper = shallow(<ExpenseList expenses={expenses} />);

    expect(wrapper).toMatchSnapshot();

});

test("should render ExpenseList without expenses", () => {

    const wrapper = shallow(<ExpenseList expenses={[]} />);

    expect(wrapper).toMatchSnapshot();

});
