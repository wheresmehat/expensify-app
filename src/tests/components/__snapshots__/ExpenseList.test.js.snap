// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="Gum bill second"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={109500}
    createdAt={-345600000}
    description="Rent first"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={4500}
    createdAt={345600000}
    description="Credit Card bill third"
    id="3"
    key="3"
    note=""
  />
</div>
`;

exports[`should render ExpenseList without expenses 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;
